File: python/python_asyncflow/kernel.py
"""AsyncKernel - Main execution engine for AsyncFlow"""

import asyncio
import time
from typing import Optional
from .generator import Generator, GeneratorBase
from .components import Node
from .time_frame import TimeFrame


class AsyncKernel(GeneratorBase, Generator):
    """Main execution kernel for the flow system"""
    
    def __init__(self, name: Optional[str] = None):
        super().__init__(name or "AsyncKernel")
        self._root = Node("Root")
        self._time_frame = TimeFrame()
        self._break_flag = False
        self._wait_until: Optional[float] = None
    
    @property
    def root(self) -> Node:
        """Get the root node"""
        return self._root
    
    @property
    def time_frame(self) -> TimeFrame:
        """Get current time frame"""
        return self._time_frame
    
    def break_flow(self) -> None:
        """Stop execution"""
        self._break_flag = True
    
    def is_breaking(self) -> bool:
        """Check if break was requested"""
        return self._break_flag
    
    def wait(self, duration: float) -> None:
        """Pause execution for duration"""
        self._wait_until = time.time() + duration
    
    def is_waiting(self) -> bool:
        """Check if currently waiting"""
        if self._wait_until is None:
            return False
        return time.time() < self._wait_until
    
    def clear_wait(self) -> None:
        """Clear wait state"""
        self._wait_until = None
    
    async def update(self, delta_time: float) -> None:
        """Update kernel with time delta"""
        self._time_frame.update_with_delta(delta_time)
        await self.step()
    
    async def update_real_time(self) -> None:
        """Update kernel with real time"""
        self._time_frame.update()
        await self.step()
    
    async def run_until_complete(self) -> None:
        """Run until all tasks complete"""
        while self.is_running() and not self.is_breaking():
            if self.is_waiting():
                await asyncio.sleep(0.001)
                continue
            
            await self.update_real_time()
            
            if self._root.child_count() == 0:
                break
            
            await asyncio.sleep(0.001)
    
    async def run_for(self, duration: float) -> None:
        """Run for specified duration"""
        start_time = time.time()
        
        while self.is_running() and not self.is_breaking():
            if time.time() - start_time >= duration:
                break
            
            if self.is_waiting():
                await asyncio.sleep(0.001)
                continue
            
            await self.update_real_time()
            await asyncio.sleep(0.001)
    
    async def step(self) -> None:
        """Execute one step of the kernel"""
        if not self.is_active() or not self.is_running() or self.is_completed():
            return
        
        if self.is_breaking():
            return
        
        if self.is_waiting():
            return
        
        child_count = self._root.child_count()
        if child_count > 0:
            self.logger.verbose(4, f"Stepping kernel with {child_count} root children")
        
        await self._root.step()
        self._root.clear_completed()

File: src/kernel.rs
use async_trait::async_trait;
use std::sync::Arc;
use std::time::Duration;
use tokio::sync::RwLock;
use tokio::time::{sleep, Instant};
use uuid::Uuid;
use crate::flow::{Generator, GeneratorBase, Node};
use crate::{Logger, TimeFrame, Result};

#[derive(Clone)]
pub struct AsyncKernel {
    base: GeneratorBase,
    root: Arc<Node>,
    time_frame: Arc<RwLock<TimeFrame>>,
    break_flag: Arc<RwLock<bool>>,
    wait_until: Arc<RwLock<Option<Instant>>>,
}

impl AsyncKernel {
    pub fn new() -> Self {
        Self {
            base: GeneratorBase::with_name("AsyncKernel"),
            root: Arc::new(Node::with_name("Root")),
            time_frame: Arc::new(RwLock::new(TimeFrame::new())),
            break_flag: Arc::new(RwLock::new(false)),
            wait_until: Arc::new(RwLock::new(None)),
        }
    }

    pub fn root(&self) -> Arc<Node> {
        self.root.clone()
    }

    pub async fn time_frame(&self) -> TimeFrame {
        let time_frame = self.time_frame.read().await;
        time_frame.clone()
    }

    pub async fn break_flow(&self) {
        let mut break_flag = self.break_flag.write().await;
        *break_flag = true;
    }

    pub async fn is_breaking(&self) -> bool {
        let break_flag = self.break_flag.read().await;
        *break_flag
    }

    pub async fn wait(&self, duration: Duration) {
        let mut wait_until = self.wait_until.write().await;
        *wait_until = Some(Instant::now() + duration);
    }

    pub async fn is_waiting(&self) -> bool {
        let wait_until = self.wait_until.read().await;
        if let Some(until) = *wait_until {
            Instant::now() < until
        } else {
            false
        }
    }

    pub async fn clear_wait(&self) {
        let mut wait_until = self.wait_until.write().await;
        *wait_until = None;
    }

    pub async fn update(&self, delta_time: Duration) -> Result<()> {
        {
            let mut time_frame = self.time_frame.write().await;
            time_frame.update_with_delta(delta_time);
        }

        self.step().await
    }

    pub async fn update_real_time(&self) -> Result<()> {
        {
            let mut time_frame = self.time_frame.write().await;
            time_frame.update();
        }

        self.step().await
    }

    pub async fn run_until_complete(&self) -> Result<()> {
        while self.is_running() && !self.is_breaking().await {
            if self.is_waiting().await {
                sleep(Duration::from_millis(1)).await;
                continue;
            }

            self.update_real_time().await?;
            
            if self.root.child_count().await == 0 {
                break;
            }

            sleep(Duration::from_millis(1)).await;
        }
        
        Ok(())
    }

    pub async fn run_for(&self, duration: Duration) -> Result<()> {
        let start_time = Instant::now();
        
        while self.is_running() && !self.is_breaking().await {
            if start_time.elapsed() >= duration {
                break;
            }

            if self.is_waiting().await {
                sleep(Duration::from_millis(1)).await;
                continue;
            }

            self.update_real_time().await?;
            sleep(Duration::from_millis(1)).await;
        }
        
        Ok(())
    }
}

impl Default for AsyncKernel {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl Generator for AsyncKernel {
    fn id(&self) -> Uuid {
        self.base.id()
    }

    fn name(&self) -> Option<&str> {
        self.base.name()
    }

    fn set_name(&mut self, name: String) {
        self.base.set_name(name);
    }

    fn is_active(&self) -> bool {
        self.base.is_active()
    }

    fn is_running(&self) -> bool {
        self.base.is_running()
    }

    fn is_completed(&self) -> bool {
        self.base.is_completed()
    }

    fn activate(&self) {
        self.base.activate();
    }

    fn deactivate(&self) {
        self.base.deactivate();
    }

    fn complete(&self) {
        self.base.complete();
    }

    async fn step(&self) -> Result<()> {
        if !self.is_active() || !self.is_running() || self.is_completed() {
            return Ok(());
        }

        if self.is_breaking().await {
            return Ok(());
        }

        if self.is_waiting().await {
            return Ok(());
        }

        let child_count = self.root.child_count().await;
        if child_count > 0 {
            self.logger().verbose(4, format!("Stepping kernel with {} root children", child_count));
        }

        self.root.step().await?;
        self.root.clear_completed().await;

        Ok(())
    }

    fn logger(&self) -> &Logger {
        self.base.logger()
    }
}